Take a stab at the sync TEMPLATE branch cleanup docs (#590)
nf-core/nf-co.re: markdown/developers/sync.md

# Fixing a broken TEMPLATE branch

If merge conflicts are resolved via the GitHub interface instead of after pulling changes to a fork as described above, the commit history from the `dev` branch will be merged into `TEMPLATE`.
This leads to complex problems in later `TEMPLATE` merges as the later updated `TEMPLATE` branch removes all the pipeline-specific files that were accidentality included in problematic merge, resulting in many (in some cases >100!) of files to resolve conflicts in.

If during one of the automated syncs you see you have an usually high number of changed files you can check whether `dev` was accidentally merged into `TEMPLATE` by looking at how many commits the `TEMPLATE` branch has (should be in the range of 5-15ish, depending on how many template updates your pipeline has had). You can also look at your repository's GitHub Network Graph under the _"Insights"_ tab or even look through the `TEMPLATE` branch commit history to see if there is a commit described as 'Merge branch `dev` into `TEMPLATE`'.

If so, the easiest solution is to start your `TEMPLATE` branch from scratch.

> * Clone the main nf-core pipeline repository to your local machine (not your development fork)

Reviewer: That's a good clarification

  ```bash

>   git clone <git URL>

Reviewer: why not including the link directly: `https://github.com/nf-core/<pipeline>.git`

Reviewer: and do we need to use https or ssh, or both are working?

Author: That would be better!

Author: ```suggestion
  git clone git@github.com:nf-core/<PIPELINE>.git
```

Author: > and do we need to use https or ssh, or both are working?

Both should work, though https may need password on the cli. I think ssh is probably "better"

  cd <pipeline>
  ```

* Next, retrieve the commit hash when the original nf-core template was used to generate pipeline i.e. with `nf-core create`.
  * Assuming you originally started with the nf-core template, you can simply look at your git log from within your repository:

    ```bash

>     git log --reverse

Reviewer: - On which branch should we do that? (I'm assuming `TEMPLATE`

Author: Shouldn't matter here actually as all branches should have the same root commit. But it will matter when we reset in the later command!

Author: ```suggestion
    git checkout TEMPLATE
    git log --reverse
```

    ```

>   * The first commit will then typically represent the original template.

Reviewer: Is there a commit message created when we use nf-core create, or do we actually suggest one like `Initial template commit`

Author: ```suggestion
  * The first commit will then typically represent the original template, with a commit message like `initial template build from nf-core/tools, version 1.9`
```

Author: `nf-core create` usually makes the first commit after version 1.9 or so..

* Reset the `TEMPLATE` branch back to this commit, discarding all changes after that

  ```bash

>   git reset --hard <hash of first commit after nf-core create>

Author: ```suggestion
    git checkout TEMPLATE
    git reset --hard <hash of first commit after nf-core create>
```

>   git reset --hard <hash of first commit after nf-core create>

Author: ```suggestion
  # Make sure you're definitely have TEMPLATE checked out!
  git reset --hard <hash of first commit after nf-core create>
```

  ```

* Push this cleaned branch back to the repository - use `--force` to overwrite the history there:

  ```bash
  git push origin TEMPLATE --force
  ```

  * This will then replace the broken `TEMPLATE` branch on GitHub with a nice clean one, which can be viewable by checking the commit history.
  * :warning: You are irreversibly overwriting git history here - make sure that you get the branch names right!
* We can switch back to `dev`, and run `nf-core sync` to do it's magic and get the latest version of the template.

  ```bash
  git checkout dev
  nf-core sync .
  ```

* You probably want to now delete your local copy of the pipeline that you checked out from the main nf-core repository.
* On your personal fork of the pipeline you'll want to pull in this fresh template branch:

  ```bash
  cd <path/to/forked/pipeline>
  git branch -D TEMPLATE # Delete the TEMPLATE branch in your fork if you have it

>   git remote add upstream <git URL of main nf-core repo> # You might already have this set up?

Author: ```suggestion
  git remote add upstream git@github.com:nf-core/<PIPELINE>.git  # You might already have this set up?
```

* Clone the main nf-core pipeline repository to your local machine (not your development fork)

  ```bash

>   git clone git@github.com:nf-core/<PIPELINE>.git

Reviewer 2: Doesn't this assume someone is using ssh keys though?

Reviewer 3: ```suggestion
  git clone https://github.com/nf-core/<PIPELINE>
```

Author: Do people not use ssh keys?

Author: @Zethson - missing the `.git`

```suggestion
  git clone https://github.com/nf-core/<PIPELINE>.git
```

Reviewer 3: @ewels very few do from my experience. The default is certainly not.

The `.git` is not required. you can clone without it.

Reviewer 2: I've seen mixed, which is why originally I think I didn't specify the exact clone mechanism.

Reviewer 3: I would say that the people that are somewhat new to git/Github will use https and not clone using their SSH keys. The people however that do use their SSH keys will know how to adapt the URL to their needs, but this does not hold the other way around (more or less).
